fpsd/features.py

```diff
@@ -0,0 +1,765 @@
+
+        self.execute_query_from_string(query)
+        return None
+
+    def _create_table_outgoing_cell_positions(self, num_cells=500):
+        """This method takes the first num_cells rows in
+        packet_positions and creates a temporary
+        table top_(num_cells)_outgoing_packet_positions
+        with the following format:
+
+        exampleid  | rank     |  outgoing_cell_position
+        (integer)  | (bigint) |  (bigint)
+        -----------------------------------------------
+        9          | 4        | 1
+        9          | 5        | 2
+        9          | 15       | 3
+
+        For example, the third outgoing cell was the 15th in the trace.
+        This table is used by the position-based feature
+        generation functions.
+        """
+        table_name = "top_{}_outgoing_packet_positions".format(num_cells)
+        self.drop_stale_feature_table(table_name)
+
+        query = ("CREATE TEMP TABLE top_{n}_outgoing_packet_positions AS "
+                 "(SELECT exampleid, outgoing_cell_position, rank        "
+                 "  FROM (                                               "
+                 "    SELECT                                             "
+                 "      ROW_NUMBER() OVER                                "
+                 "      (PARTITION BY exampleid ORDER BY rank)           "
+                 "      AS outgoing_cell_position,                       "
+                 "      t.*                                              "
+                 "    FROM packet_positions t) x                         "
+                 "  WHERE x.outgoing_cell_position <= {n}                "
+                 ");").format(n=num_cells)
+
+        self.execute_query_from_string(query)
+        return None
+
+    def get_exampleids(self):
+        """Get list of exampleids"""
+        query = "SELECT DISTINCT exampleid FROM raw.frontpage_traces"
+        df = pd.read_sql(query, self.engine)
+        return df.exampleid.values
+
+    def get_trace_cells(self, exampleid):
+        """Get trace for a given exampleid"""
+        df = pd.read_sql(("SELECT ingoing, t_trace FROM raw.frontpage_traces "
+                          "WHERE exampleid={} "
+                          "ORDER BY t_trace").format(exampleid),
+                         self.engine)
+        return df
+
+    def generate_table_cell_numbers(self):
+        """This method takes all examples and produces a
+        table features.cell_numbers with the following format:
+
+        exampleid  | a        | b        | c
+        (integer)  | (bigint) | (bigint) | (bigint)
+        -----------------------------------------------
+        9          | 326      | 37       | 363
+        10         | 493      | 47       | 540
+        11         | 652      | 77       | 729
+
+        where a, b, c are the feature columns:
+
+        total_number_of_incoming_cells,
+        total_number_of_outgoing_cells,
+        and total_number_of_cells.
```

Comment:
I _can_ put it here, but am going to leave as is because these two tables are logically different classes of features - one contains packet numbers and one contains timing - so it seems reasonable to leave them separate. Note that when everything is merged together in the master view in the end the ordering of columns is as you suggest here